letsencrypt-auto-source/rebuild_dependencies.py

```diff
@@ -0,0 +1,221 @@
+
+        with open(os.path.join(workspace, 'results'), 'r') as file_h:
+            return file_h.read()
+    finally:
+        if os.path.isfile(cid_file):
+            with open(cid_file, 'r') as file_h:
+                cid = file_h.read()
+            subprocess.call(['docker', 'kill', cid], stdout=None, stderr=None)
+        shutil.rmtree(workspace)
+
+
+def _insert_results(dependencies_map, results, distribution):
+    refined_results = []
+    for result in results.split(os.linesep):
+        match = re.match(r'(.*)==(.*)', result)
+        if match:
+            package = match.group(1)
+            version = match.group(2)
+            if not any(dep in package for dep in ['acme', 'certbot', 'pkg-resources']):
+                dependencies_map.setdefault(package, []).append((version, distribution))
+
+    return refined_results
+
+
+def _process_dependency_map(dependency_map):
+    print('===> Processing the dependency map.')
+    requirements = []
+    conflicts = []
+    for package, versions in dependency_map.items():
+        reduced_versions = _reduce_versions(versions)
+
+        if len(reduced_versions) > 1:
+            version_list = ['{0} ({1})'.format(version, ','.join(distributions))
+                            for version, distributions in reduced_versions.items()]
+            conflict = ('package {0} is declared with several versions: {1}'
+                        .format(package, ', '.join(version_list)))
+            conflicts.append(conflict)
+            sys.stderr.write('ERROR: {0}\n'.format(conflict))
+        else:
+            requirements.append((package, list(reduced_versions)[0]))
+
+    requirements.sort(key=lambda x: x[0])
+    return requirements, conflicts
+
+
+def _reduce_versions(versions):
+    version_map = {}
+    for version in versions:
+        version_map.setdefault(version[0], []).append(version[1])
+
+    return version_map
+
+
+def _write_requirements(dest_file, requirements, conflicts):
+    print('===> Calculating hashes for the requirement file.')
+    if os.path.exists(dest_file):
+        os.remove(dest_file)
+
+    with open(dest_file, 'w') as file_h:
+        file_h.write('''\
+# This is the flattened list of packages certbot-auto installs.
+# To generate this, do (with docker and package hashin installed):
+# ```
+# letsencrypt-auto-source/rebuild_dependencies.py \\
+#   letsencrypt-auto-sources/pieces/dependency-requirements.txt
+# ```
+''')
+    for req in requirements:
+        subprocess.check_call(['hashin', '{0}=={1}'.format(req[0], req[1]),
+                               '--requirements-file', dest_file])
+
+    if conflicts:
+        with open(dest_file, 'a') as file_h:
+            file_h.write('\n## ! SOME ERRORS OCCURRED ! ##\n')
+            file_h.write('\n'.join('# {0}'.format(conflict) for conflict in conflicts))
+            file_h.write('\n')
+
+    with open(dest_file, 'r') as file_h:
+        print(file_h.read())
```

Comment:
Perhaps it's not important for this program yet, but I generally find it advantageous to keep side effects (like non-debug prints) out of internal functions. It allows for reuse and easy testing. The same applies to the print at the end of _gather_dependencies().

Reply:
Done.